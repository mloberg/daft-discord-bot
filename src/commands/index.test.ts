import { Client, Message, TextChannel } from 'discord.js';

import { FriendlyError } from '../error';
import { Commands, help } from '.';

const mocks = {
    send: jest.fn(),
};

jest.mock('discord.js', () => ({
    Message: jest.fn().mockImplementation(() => ({
        channel: {
            send: mocks.send,
        },
    })),
}));

describe('_help configuration', () => {
    it('should have basic command infomation', () => {
        expect(help.name).toEqual('help');
        expect(help.description).toEqual('Get help with commands');
        expect(help.usage).toEqual('[COMMAND]');
    });

    it('should have an alias', () => {
        expect(help.alias).toContain('commands');
    });
});

describe('_help', () => {
    let message: Message;

    beforeEach(() => {
        mocks.send.mockClear();
        mocks.send.mockReturnThis();

        const client = {} as Client;
        const channel = {} as TextChannel;
        message = new Message(client, {}, channel);
    });

    it('returns a list of all commands', async () => {
        const reply = await help.run(message, { _: [], $0: 'help' });

        expect(reply).toEqual(message.channel);
        expect(mocks.send).toMatchSnapshot();
    });

    it.each(['help', 'add', 'play', 'pause', 'volume'])('returns the details for the %s command', async (cmd) => {
        const reply = await help.run(message, { _: [cmd], $0: 'help' });

        expect(reply).toEqual(message.channel);
        expect(mocks.send).toMatchSnapshot();
    });

    it('returns an error if an invalid command is given', async () => {
        try {
            await help.run(message, { _: ['invalid'], $0: 'help' });

            fail('expected error to be thrown');
        } catch (err) {
            expect(err instanceof FriendlyError).toBe(true);
            expect(err.message).toEqual('No command for "invalid" found.');
        }
    });
});

describe('Commands', () => {
    it('registers commands and their aliases', () => {
        const commands = new Commands();
        expect(commands.list).toHaveLength(0);

        commands.register(help);
        expect(commands.list).toHaveLength(1);
        expect(commands.all).toHaveLength(2);
    });

    it('returns the correct command', () => {
        const commands = new Commands();
        commands.register(help);

        expect(commands.get('help')).toEqual(help);
        expect(commands.get('commands')).toEqual(help);
        expect(commands.get('foo')).toBeNull();
    });
});
