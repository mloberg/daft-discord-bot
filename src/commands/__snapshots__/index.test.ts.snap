// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_help returns a list of all commands 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "Here is a list of available commands:",
        "help, add, manage, next, pause, ping, play, playing, resume, stop, tag, volume",
        "Get more details with \\"_help [command]\\"",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "Here is a list of available commands:",
                "help, add, manage, next, pause, ping, play, playing, resume, stop, tag, volume",
                "Get more details with \\"_help [command]\\"",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the add command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "**add**: Add a song",
        "*usage*: \`_add [FILE] [...TAGS]\`",
        "*examples*: \`_add counting-the-cost.webm dark menacing epic\`",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "**add**: Add a song",
                "*usage*: \`_add [FILE] [...TAGS]\`",
                "*examples*: \`_add counting-the-cost.webm dark menacing epic\`",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the help command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "**help**: Get help with commands",
        "*aliases*: commands",
        "*usage*: \`_help [COMMAND]\`",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "**help**: Get help with commands",
                "*aliases*: commands",
                "*usage*: \`_help [COMMAND]\`",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the pause command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "**pause**: Pause the playlist",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "**pause**: Pause the playlist",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the play command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "**play**: Start a playlist",
        "*aliases*: start, playlist",
        "*usage*: \`_play ...TAGS|SONG [--volume|-v VOLUME]\`",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "**play**: Start a playlist",
                "*aliases*: start, playlist",
                "*usage*: \`_play ...TAGS|SONG [--volume|-v VOLUME]\`",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;

exports[`_help returns the details for the volume command 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Array [
        "**volume**: Set the playback volume",
        "*usage*: \`_volume [VOLUME]\`",
      ],
      Object {
        "split": true,
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "send": [MockFunction] {
          "calls": Array [
            Array [
              Array [
                "**volume**: Set the playback volume",
                "*usage*: \`_volume [VOLUME]\`",
              ],
              Object {
                "split": true,
              },
            ],
          ],
          "results": [Circular],
        },
      },
    },
  ],
}
`;
